src/internals/ir/callGraph.ts

```diff
@@ -0,0 +1,193 @@
+import { TactASTStore } from "./astStore";
+import { IdxGenerator } from "./indices";
+import {
+  AstFunctionDef,
+  AstReceiver,
+  AstContractInit,
+  AstStatement,
+  AstStaticCall,
+  AstMethodCall,
+  AstExpression,
+  AstStructFieldInitializer,
+} from "@tact-lang/compiler/dist/grammar/ast";
+
+class CGEdge {
+  public idx: number;
+  constructor(
+    public src: number,
+    public dst: number,
+  ) {
+    this.idx = IdxGenerator.next("edge");
+  }
+}
+
+class CGNode {
+  public idx: number;
+  constructor(
+    public astId: number,
+    public name: string,
+  ) {
+    this.idx = IdxGenerator.next("node");
+  }
+}
+
+export class CallGraph {
+  public nodes: CGNode[] = [];
+  public edges: CGEdge[] = [];
+  private nodeMap: Map<number, CGNode> = new Map();
+  private edgeMap: Map<number, Set<number>> = new Map();
+
+  constructor(public astStore: TactASTStore) {}
+
+  build() {
+    this.addFunctionsToNodes();
+    this.analyzeFunctionCalls();
+  }
+
+  private addFunctionsToNodes() {
+    for (const func of this.astStore.getFunctions()) {
+      const funcName = this.getFunctionName(func);
+      if (funcName) {
+        const node = new CGNode(func.id, funcName);
+        this.nodes.push(node);
+        this.nodeMap.set(func.id, node);
+      }
+    }
+  }
+
+  private analyzeFunctionCalls() {
+    for (const func of this.astStore.getFunctions()) {
+      const funcName = this.getFunctionName(func);
+      if (funcName) {
+        this.processStatements(func.statements, func.id);
+      }
+    }
+  }
+
+  private getFunctionName(
+    func: AstFunctionDef | AstReceiver | AstContractInit,
+  ): string | undefined {
+    if (func.kind === "function_def") return func.name.text;
+    if (func.kind === "contract_init") return "contract_init";
+    if (func.kind === "receiver") return "receiver";
+    return undefined;
```

Comment:
You could use `switch` with the `unreachable` call in the `default` clause. This will give us additional static guarantees for types of the function's arguments.